### PR TITLE
Add header tracking for Linux Makefile builds

### PR DESCRIPTION
**Pull request by MEhrn00**
This PR adds build flags for Mingw GCC to emit make dependency files. This allows make to automatically handle prerequisites on header files.

### CODE REVIEW

File: Payload_Type/hannibal/hannibal/agent_code/Hannibal/linux_makefile [hunk 1]
```diff
@@ -63,25 +65,31 @@ $(BIN_DIR)/$(PROJECT).exe: $(ASM_OBJ_FILES) $(OBJ_FILES)
...
 	@rm -rf bin/obj/*.ii
+	@rm -rf $(DEP_DIR)
 	@rm -rf bin/*.bin
 	@rm -rf bin/*.exe
 
 print:
 	@echo "SRC_FILES: $(SRC_FILES)"
 	@echo "OBJ_FILES: $(OBJ_FILES)"
 	@echo "ASM_FILES: $(ASM_FILES)"
-	@echo "ASM_OBJ_FILES": $(ASM_OBJ_FILES)"
```
Comment (silentwarble) [inline Payload_Type/hannibal/hannibal/agent_code/Hannibal/linux_makefile:87]: I'm assuming you meant to put `@echo "DEP_FILES: $(DEPFILES)" `here. 

I also need to remove the extra ": `@echo "ASM_OBJ_FILES" <---- : $(ASM_OBJ_FILES)"`

**Review by silentwarble — CHANGES_REQUESTED** [review]
Hey thanks a ton for this! See my one comment on @echo "DEP_FILES: $(DEP_FILES)"  and removing an extra " that I had in another echo statement. Other than that everything worked when I tested.

File: Payload_Type/hannibal/hannibal/agent_code/Hannibal/linux_makefile (hunk 1 above)
Comment (silentwarble) [reply Payload_Type/hannibal/hannibal/agent_code/Hannibal/linux_makefile:87]: Doh, I meant `@echo "DEP_FILES: $(DEP_FILES)" `

Comment (MEhrn00) [reply Payload_Type/hannibal/hannibal/agent_code/Hannibal/linux_makefile:87]: I added the `@echo "DEP_FILES: $(DEP_FILES)"` line in the `print` target here 3b4672d6e194285bc25340b849b0db1df981ebc1.

Comment (MEhrn00) [reply Payload_Type/hannibal/hannibal/agent_code/Hannibal/linux_makefile:87]: Removed the extraneous quote for the `ASM_OBJ_FILES` 6653196862c35d9ce835d4af2e385d195ad549ef.

**Comment by MEhrn00** [comment]
I added those changes.

Truthfully, I did not know what the `print` target was used for so I decided to leave it untouched. It may be a good idea to mark the all, clean and print targets as [`.PHONY`](https://www.gnu.org/software/make/manual/html_node/Phony-Targets.html) since they do not produce any side effects.

This should also help solve some of the other build quirks such as the final executable link target removing all of the object files at the end forcing an entire rebuild. With the headers being added as prereqs, changes to the config should force a correct rebuild of only the necessary object files.
https://github.com/MythicAgents/Hannibal/blob/be0855db49ca3723eee68f1fce649d32240aafc3/Payload_Type/hannibal/hannibal/agent_code/Hannibal/linux_makefile#L59-L64

**Comment by silentwarble** [comment]
Yeah I was experimenting with that. I think I can remove them now. Thanks for your contribution!

**Pull request merged**